### DISCUSSION TITLE
correct way to use fzf binary, key-bindings and completion with zinit

### DISCUSSION THREAD

**Discussion by francois-pasquier** [zdharma-continuum/zinit#359] [Q&A] [answered]
For the moment, I managed to get it working with this snippet:

```zsh
zi ice wait lucid from"gh-r" as"program" mv"fzf* -> fzf" pick"fzf/fzf"
zi light junegunn/fzf-bin

zinit ice wait lucid from"gh-r" nocompile src'shell/key-bindings.zsh' sbin \
      dl'https://raw.githubusercontent.com/junegunn/fzf/master/shell/completion.zsh -> _fzf_completion;
         https://raw.githubusercontent.com/junegunn/fzf/master/shell/key-bindings.zsh -> key-bindings.zsh;
         https://raw.githubusercontent.com/junegunn/fzf/master/man/man1/fzf-tmux.1 -> $ZPFX/share/man/man1/fzf-tmux.1;
         https://raw.githubusercontent.com/junegunn/fzf/master/man/man1/fzf.1 -> $ZPFX/share/man/man1/fzf.1'
zinit light junegunn/fzf
```

Is there any shorter way to do it?
I'd also like to skip using whole urls and simply using the content of `junegunn/fzf` plugin that I could install with `zi light junegunn/fzf`

Honestly, it felt like quite an achievement to get fzf working as I didn't find any documentation to do so (outside of just the binary itself)

**Comment by vladdoster** [comment, marked as answer]
I condensed the URLs using brace expansion (which can be optimized further).

```zsh
zi for \
    as'command' \
    dl="$(print -c https://raw.githubusercontent.com/junegunn/fzf/master/{shell/{'key-bindings.zsh;','completion.zsh -> _fzf;'},man/{'man1/fzf.1 -> $ZPFX/share/man/man1/fzf.1;','man1/fzf-tmux.1 -> $ZPFX/share/man/man1/fzf-tmux.1;'}})" \
    from'gh-r' \
    null \
    pick'fzf' \
    src'key-bindings.zsh' \
  @junegunn/fzf
```

**Reply by vladdoster** [reply]
Really depends if you are ok with a long line (less readable) that is harder to grok or fewer lines of code :).

**Reply by ngocphamm** [reply]
This is great @vladdoster. However I'm still having trouble understanding zinit commands.

Could you possibly explain a few things about the above command for me please?

1. What is the `null` (5th line) for?
2. I understand that we have to source `key-binding.zsh`, move man files for `fzf` and `fzf-tmux` to the correct folder inside `$ZPFX`. But why don't we have to do anything with the completion file, besides renaming it to `_fzf`?
3. Why do we need `@` in front of `junegunn/fzf`? Would it colide with any other ice like we have to do with `@sharkdp`, for example, because `sh` is an ice?

**Reply by jankatins** [reply]
1.) I was curiose about the null (which seems to be not documented): if I read this line right, it's a shorthand for `pick'/dev/null'`, so essentially "don't source anything" (there are two more places with `ICE[null]`, both disable loading snippets/plugins as well.
2.) The default for as"command" ([but not `as"null"`!](https://github.com/zdharma-continuum/zinit/pull/417)) is to use everything which has [an underscore at the front as a completion (=link it to the completion folder which gets loaded at zinit start)](https://github.com/zdharma-continuum/zinit/blob/59975d70d7101651b0bb01f4e22c23db0dae8133/zinit-install.zsh#L553-L586).
3.) No idea if it is needed, but I also started adding it to all my github slugs to not get strange errors when it would be needed 8like in the sharkdb case)...

**Reply by ngocphamm** [reply]
@jankatins 
1. But we have `pick'fzf'` right there too 🤔 
2. That's it. Thank you!
3. That's actually a good idea. I think that can even be enforced, like `@` will be like a "pointer" of where we want to get the source from.

**Reply by ngocphamm** [reply]
And I forgot to ask about the `dl` ice too. I tried but couldn't find it  in any of the documentation. What exactly it does?

EDIT: Nevermind. Found it. https://github.com/zdharma-continuum/zinit-annex-patch-dl

**Reply by ngocphamm** [reply]
So I've been playing around and this works for me to get all the `fzf` related stuff that I need, including the `fzf-tmux` binary, using `sbin` ice, so cleaner `$PATH`.

The only thing I still need to figure out is the completion, because I use `as"null"` there.

```
zinit as"null" wait"0a" lucid light-mode from"gh-r" for \
  sbin"fzf*" \
    dl'https://raw.githubusercontent.com/junegunn/fzf/master/bin/fzf-tmux;
       https://raw.githubusercontent.com/junegunn/fzf/master/shell/completion.zsh -> _fzf_completion;
       https://raw.githubusercontent.com/junegunn/fzf/master/shell/key-bindings.zsh;
       https://raw.githubusercontent.com/junegunn/fzf/master/man/man1/fzf-tmux.1 -> $ZPFX/man/man1/fzf-tmux.1;
       https://raw.githubusercontent.com/junegunn/fzf/master/man/man1/fzf.1 -> $ZPFX/man/man1/fzf.1' \
    src"key-bindings.zsh" \
    junegunn/fzf
```

And yeah I prefer multiple lines over one long line. 🙃

**Reply by ngocphamm** [reply]
Another behavior I'm confused about is that in those `dl` ice lines, I can use `$ZPFX` but not those `$ZINIT[...]` https://github.com/zdharma-continuum/zinit#customizing-paths

I tried to point the man file to `$ZINIT[MAN_DIR]`, or the completion file to `$ZINIT[COMPLETIONS_DIR]` but a folder named `$ZINIT[COMPLETIONS_DIR]` is created, with the `_fzf_completion` file in it, instead.

`atclone'echo $ZPFXl; echo $ZINIT[COMPLETIONS_DIR];'` still can print the directories correctly though 🤔 

```
Downloading junegunn/fzf...
(Requesting `fzf-0.35.1-darwin_amd64.zip'...)
####################################################################################################################################################################################################################################### 100.0%
[ziextract] Unpacking the files from: `fzf-0.35.1-darwin_amd64.zip'...
Archive:  fzf-0.35.1-darwin_amd64.zip
  inflating: fzf
[ziextract] Successfully extracted and assigned +x chmod to the file: fzf.
[patch-dl annex]: File fzf-tmux downloaded correctly
[patch-dl annex]: File _fzf_completion downloaded correctly
[patch-dl annex]: File key-bindings.zsh downloaded correctly
[patch-dl annex]: File fzf-tmux.1 downloaded correctly
[patch-dl annex]: File fzf.1 downloaded correctly
/Users/ngoc/.local/share/zinit/polaris
/Users/ngoc/.local/share/zinit/completions
bin-gem-node annex: Created the fzf shim and set +x on the fzf binary
bin-gem-node annex: Created the fzf-tmux shim and set +x on the fzf-tmux binary
```

**Reply by jankatins** [reply]
With latest HEAD, you should be able to use the "completion" ice to overwrite the behavior of `as"null"`.

**Reply by ngocphamm** [reply]
Thank you! @jankatins So I've been looking at a new feature added just days ago, without having latest version. I thought I did `zi update` though.

Now `_fzf_completion` is being symlinked properly to `$ZINIT[COMPLETIONS_DIR]`.

However, I think there's still something special with this. Normally, FZF would want to source the completion like this, and after that the function `fzf-completion` would be available.

```
[[ $- == *i* ]] && source "/usr/local/opt/fzf/shell/completion.zsh" 2> /dev/null
```

With `zinit`, even though `zi clist | grep fzf` shows that the plugin's completion has been loaded, `type fzf-completion` returns `not found`. 🤔

**Reply by jankatins** [reply]
I can reproduce, but have no idea why this is so... @vladdoster Maybe your debugging skills get this further...

(also found out that `zinit completions for ...` will print the completions, but `zinit lbin"!" completions for ...` will do the right thing. Not sure if thats a bug or a side effect...

**Reply by vladdoster** [reply]
@jankatins yeah I noticed the behavior of `zinit completions for ...` which makes sense since `completions` is a valid sub command.

**Comment by fky2015** [comment]
This plugin also does a wonderful job: https://github.com/joshskidmore/zsh-fzf-history-search
